design/kanister-kopia-integration/kanister-kopia-integration.md

```diff
@@ -0,0 +1,91 @@
+
+Kopia is a powerful, cross-platform tool for managing encrypted backups in the cloud.
+It provides fast and secure backups, using compression, data deduplication, and client-side end-to-end encryption.
+It supports a variety of backup storage targets, including object stores, which allows users to choose the storage provider that better addresses their needs. In Kopia, these storage locations are called repositories.
+It is a lock-free system that allows concurrent multi-client operations including garbage collection.
+
+To explore other features of Kopia, see its [documentation](https://kopia.io/docs/features/).
+
+## Scope
+
+1. Automate the initialization of a Kopia Repository for an application backed up by Kanister.
+2. Design and automate the lifecycle of the required Kopia Repository Server.
+3. Add new versions of Kanister Data Functions like BackupData, RestoreData, etc. with Kopia as the primary data mover tool.
+
+## User Experience
+
+- All the new features mentioned in this document will be opt-in only. Existing users will not see any changes in the Kanister controller's behavior.
+- Users will be able to continue using their current Blueprints, switch to the v2 version of the example Blueprints, or use Blueprints with the new version of the Kanister Data Functions.
+- Users opting to use the v2 Blueprints and Blueprints with Kopia-based Kanister Data Functions will be required to follow instructions to set up the required Kopia Repository Server before executing the actions.
+- After setting up the Repository Server, users can follow the normal workflow to execute actions from the v2 Blueprints. To use the new versions of the Kanister Data Functions, users must specify the version of the function via the ActionSet Action's `preferredVersion` field.
+
+## Detailed Design
+
+### Kopia Repository
+
+- As mentioned above, the backup storage location is called a "Repository" in Kopia.
+- A separate repository will be used for each application protected by Kopia-based Blueprints in Kanister.
+- The repository will be initialized when the Repository Server is created the first time. Once initialized, it will continue to exist at the location until cleaned up manually.
+- Accessing the repository requires location and credential information similar to a Kanister Profile CR and a unique password used by Kopia during [encryption](https://kopia.io/docs/features/#end-to-end-zero-knowledge-encryption).
+- In the first iteration, this password will be auto-generated by the Kanister controller. Future iterations will allow users to use a Key Management Service of choice.
+- Only a single repository can exist at a particular backend storage location. To address this, the Kanister controller will generate the repository path using the location information and the UUID of the application namespace. For example, the repository path for `mysql` namespace and S3 bucket called `test-bucket` will be of the form `test-bucket/<UUID of mysql namespace>`.
```

Comment:
`Only a single repository can exist at a particular backend storage location.`
in the second point here, we are saying that we are going to create kopia repo (repo) for every application and in this sentence we are saying we can only have one repo in the backend storage location.
That seems a bit off? Because it would mean we can just backup one application in one backup storage location.

Reply:
do we really need bucket name in the repo path?
Bucket name is already part of the object URL I think.

Reply:
I have updated it to clarify the limitation. Yes, it is true that only a single repository can exist at a particular path which is why we need to add a unique string in the path to allow creation of multiple repos in the same bucket.

What do you mean by object URL? How can that be used in a generated string?

Reply:
so let's say we have an object in the bucket  `infracloud.kanister.io` at the path `abc/def/a.txt`, the path to this object is going to look like 
```
s3://infracloud.kanister.io/abc/def/a.txt
```
but if we decide that the file is going to be present at path `<bucket>/abc/def/a.txt` in that case the path to object is going to look like
```
s3://<bucket>/<bucket>/abc/def/a.txt
```
I am not sure but I think is what going to happen.

Reply:
That's not right. If the path provided by the user is `s3://<bucket>/` then the path to the repository will be `s3://<bucket>/<application namespace UID>`